pkg/controllers/provisioning/static/controller.go

```diff
@@ -0,0 +1,146 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package static
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+	"sigs.k8s.io/karpenter/pkg/utils/resources"
+
+	"sigs.k8s.io/karpenter/pkg/controllers/provisioning"
+	"sigs.k8s.io/karpenter/pkg/controllers/state"
+	"sigs.k8s.io/karpenter/pkg/events"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+
+	"sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	nodepoolutils "sigs.k8s.io/karpenter/pkg/utils/nodepool"
+)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cluster       *state.Cluster
+	ncProvisioner *provisioning.NCProvisioner
+}
+
+func NewController(kubeClient client.Client, cluster *state.Cluster, recorder events.Recorder, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		cluster:       cluster,
+		ncProvisioner: provisioning.NewNCProvisioner(kubeClient, recorder, cluster),
+	}
+}
+
+// Reconcile the resource
+func (c *Controller) Reconcile(ctx context.Context, np *v1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "provisioning.static")
+	// TODO: This can be improved to not wait for cluster sync but we need to either rely on a manual
+	// update of NodePoolResources so that we track node count accurately, or handle getting NodeClaim count some other way
+	if !c.cluster.Synced(ctx) {
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
+	if !nodepoolutils.IsManaged(np, c.cloudProvider) {
+		return reconcile.Result{}, nil
+	}
+	if !np.StatusConditions().Root().IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	if np.Spec.Replicas == nil {
+		return reconcile.Result{}, nil
+	}
+	nodes := c.cluster.NodePoolResourcesFor(np.Name)[resources.Node]
+	// Size down of replicas will be handled in disruption controller to drain nodes and delete NodeClaims
+	if nodes.Value() > lo.FromPtr(np.Spec.Replicas) {
+		return reconcile.Result{}, nil
+	}
+	// Create the number of NodeClaims equal to the desired replica count
+	var nodeClaims []*scheduling.NodeClaim
+	for range lo.FromPtr(np.Spec.Replicas) - nodes.Value() {
+		// Need fresh copy everytime else nodeclaims would share the template resources causing races
+		nct := scheduling.NewNodeClaimTemplate(np)
```

Comment:
Can we inject the instance types that we are actually resolving to here?

Reply:
done